stdlib/internal/SwiftSyntax/SyntaxCollection.swift

```diff
@@ -0,0 +1,118 @@
+    _ layout: [RawSyntax]) -> SyntaxCollection<SyntaxElement> {
+    let newRaw = data.raw.replacingLayout(layout)
+    let (newRoot, newData) = data.replacingSelf(newRaw)
+    return SyntaxCollection<SyntaxElement>(root: newRoot, data: newData)
+  }
+
+  /// Creates a new SyntaxCollection by appending the provided syntax element
+  /// to the children.
+  ///
+  /// - Parameter syntax: The element to append.
+  /// - Returns: A new SyntaxCollection with that element appended to the end.
+  public func appending(
+    _ syntax: SyntaxElement) -> SyntaxCollection<SyntaxElement> {
+    var newLayout = data.raw.layout
+    newLayout.append(syntax.raw)
+    return replacingLayout(newLayout)
+  }
+
+  /// Creates a new SyntaxCollection by prepending the provided syntax element
+  /// to the children.
+  ///
+  /// - Parameter syntax: The element to prepend.
+  /// - Returns: A new SyntaxCollection with that element prepended to the
+  ///            beginning.
+  public func prepending(
+    _ syntax: SyntaxElement) -> SyntaxCollection<SyntaxElement> {
+    return inserting(syntax, at: 0)
+  }
+
+  /// Creates a new SyntaxCollection by inserting the provided syntax element
+  /// at the provided index in the children.
+  ///
+  /// - Parameters:
+  ///   - syntax: The element to insert.
+  ///   - index: The index at which to insert the element in the collection.
+  ///
+  /// - Returns: A new SyntaxCollection with that element appended to the end.
+  public func inserting(_ syntax: SyntaxElement,
+                        at index: Int) -> SyntaxCollection<SyntaxElement> {
+    var newLayout = data.raw.layout
+    newLayout.insert(syntax.raw, at: index)
+    return replacingLayout(newLayout)
+  }
+
+  /// Creates a new SyntaxCollection by removing the syntax element at the
+  /// provided index.
+  ///
+  /// - Parameter index: The index of the element to remove from the collection.
+  /// - Returns: A new SyntaxCollection with the element at the provided index
+  ///            removed.
+  public func removing(childAt index: Int) -> SyntaxCollection<SyntaxElement> {
+    var newLayout = data.raw.layout
+    newLayout.remove(at: index)
+    return replacingLayout(newLayout)
+  }
+
+  /// Creates a new SyntaxCollection by removing the first element.
+  ///
+  /// - Returns: A new SyntaxCollection with the first element removed.
+  public func removingFirst() -> SyntaxCollection<SyntaxElement> {
+    var newLayout = data.raw.layout
+    newLayout.removeFirst()
+    return replacingLayout(newLayout)
+  }
+
+  /// Creates a new SyntaxCollection by removing the last element.
+  ///
+  /// - Returns: A new SyntaxCollection with the last element removed.
+  public func removingLast() -> SyntaxCollection<SyntaxElement> {
+    var newLayout = data.raw.layout
+    newLayout.removeLast()
+    return replacingLayout(newLayout)
+  }
+}
+
+/// Conformance for SyntaxCollection to the Collection protocol.
+extension SyntaxCollection: Collection {
```

Comment:
Based on the methods in the class itself, it can also conform to `MutableCollection` and `RangeReplaceableCollection` and potentially acquire a more familiar API for things like `removingLast` etc...